crypto/fipsmodule/rand/snapsafe_detect.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/crypto.h>
+
+#include "snapsafe_detect.h"
+
+#if defined(OPENSSL_LINUX)
+#include "../delocate.h"
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+/* Snapsafety state */
+#define SNAPSAFETY_STATE_FAILED_INITIALISE 0x00
+#define SNAPSAFETY_STATE_SUCCESS_INITIALISE 0x01
+#define SNAPSAFETY_STATE_NOT_SUPPORTED 0x02
+
+DEFINE_STATIC_ONCE(aws_snapsafe_init)
+DEFINE_BSS_GET(const char *, sgc_file_path)
+DEFINE_BSS_GET(volatile uint32_t *, sgc_addr)
+DEFINE_BSS_GET(int, snapsafety_state)
+
+/* aws_snapsafe_check_kernel_support returns 1 if the special sysgenid device
+ * file exists and 0 otherwise. */
+static int aws_snapsafe_check_kernel_support(void) {
+  /* This file-exist method is generally brittle. But for our purpose, this
+   * should be more than fine. */
+  if (access(*sgc_file_path_bss_get(), F_OK) != 0) {
+    return 0;
+  }
+  return 1;
+}
+
+static void do_aws_snapsafe_init(void) {
+
+#if defined(AWSLC_SYSGENID_PATH)
+  *sgc_file_path_bss_get() = AWSLC_SYSGENID_PATH;
+#else
+  *sgc_file_path_bss_get() = "/dev/sysgenid";
+#endif
+  *sgc_addr_bss_get() = NULL;
+
+  if (aws_snapsafe_check_kernel_support() != 1) {
+    *snapsafety_state_bss_get() = SNAPSAFETY_STATE_NOT_SUPPORTED;
+    return;
+  }
+
+  /* While not totally obvious from man pages, mmap actually only allocates
+   * memory on page size boundaries. So just give it that hint. */
+  long page_size = sysconf(_SC_PAGESIZE);
+  if (page_size <= 0) {
+    *snapsafety_state_bss_get() = SNAPSAFETY_STATE_FAILED_INITIALISE;
+    return;
+  }
+
+  if (*sgc_addr_bss_get() == NULL) {
+    int fd_sgc = open(*sgc_file_path_bss_get(), O_RDONLY);
+    if (fd_sgc == -1) {
+      *snapsafety_state_bss_get() = SNAPSAFETY_STATE_FAILED_INITIALISE;
+      return;
+    }
+
+    void *addr =
+        mmap(NULL, (size_t)page_size, PROT_READ, MAP_SHARED, fd_sgc, 0);
+
+    /* Can close file descriptor now per
+     * https://man7.org/linux/man-pages/man2/mmap.2.html: "After the mmap() call
+     * has returned, the file descriptor, fd, can be closed immediately without
+     * invalidating the mapping.". We have initialised snapsafety without errors
+     * and this function is only executed once. Therefore, try to close file
+     * descriptor but don't error if it fails. */
+    close(fd_sgc);
+
+    if (addr == MAP_FAILED) {
+      *snapsafety_state_bss_get() = SNAPSAFETY_STATE_FAILED_INITIALISE;
+      return;
+    }
+
+    /* sgc_addr will now point at the mapped memory and any 4-byte read from
+     * this pointer will correspond to the sgn manager by the VMM. */
+    *sgc_addr_bss_get() = addr;
+
+    *snapsafety_state_bss_get() = SNAPSAFETY_STATE_SUCCESS_INITIALISE;
+  }
+}
+
+static uint32_t aws_snapsafe_read_sgn(void) {
+  if (*snapsafety_state_bss_get() == SNAPSAFETY_STATE_SUCCESS_INITIALISE) {
+    return **sgc_addr_bss_get();
+  }
```

Comment:
Just double checking are we reading the right part of the mmap data? From the documentation:
> The driver supports ``PROT_READ, MAP_SHARED`` mmaps of a single page
> in size. The first 4 bytes of the mapped page will contain an
>  up-to-date u32 copy of the system generation counter.

So the pointer points to the page of memory and the counter is at address -> address + 3, not something like an endian issue and the counter is on the other side of the page.

Reply:
As long as we're reading the first (i.e., smallest-address) 4 bytes of the page and interpreting them as an (endian-appropriate) u32 value, the logic should be correct.